test/e2e/garbage_collector.go

```diff
@@ -441,4 +472,239 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
 		gatherMetrics(f)
 	})
 
+	It("[Feature:GarbageCollector] should keep the rc around until all its pods are deleted if the deleteOptions says so", func() {
+		clientSet := f.ClientSet
+		rcClient := clientSet.Core().ReplicationControllers(f.Namespace.Name)
+		podClient := clientSet.Core().Pods(f.Namespace.Name)
+		rcName := "simpletest.rc"
+		rc := newOwnerRC(f, rcName, 2)
+		replicas := int32(100)
+		rc.Spec.Replicas = &replicas
+		By("create the rc")
+		rc, err := rcClient.Create(rc)
+		if err != nil {
+			framework.Failf("Failed to create replication controller: %v", err)
+		}
+		// wait for rc to create pods
+		if err := wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
+			rc, err := rcClient.Get(rc.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, fmt.Errorf("Failed to get rc: %v", err)
+			}
+			if rc.Status.Replicas == *rc.Spec.Replicas {
+				return true, nil
+			} else {
+				return false, nil
+			}
+		}); err != nil {
+			framework.Failf("failed to wait for the rc.Status.Replicas to reach rc.Spec.Replicas: %v", err)
+		}
+		By("delete the rc")
+		deleteOptions := getForegroundOptions()
+		deleteOptions.Preconditions = metav1.NewUIDPreconditions(string(rc.UID))
+		if err := rcClient.Delete(rc.ObjectMeta.Name, deleteOptions); err != nil {
+			framework.Failf("failed to delete the rc: %v", err)
+		}
+		By("wait for the rc to be deleted")
+		// default client QPS is 20, deleting each pod requires 2 requests, so 30s should be enough
+		if err := wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
```

Comment:
The test is verifying that *i)* the rc is not deleted before all its 100 pods are deleted, and *ii)* that all objects are deleted within 30s.

For *i)*, the test is not strict, because the test will still pass if the rc is deleted after 99 pods are deleted, and the last pod is deleted after the rc is deleted but before the test issues `podClient.List()`. The test can be strict if it can compare the resource version of the last pod that gets deleted and the resource version of the rc when it gets deleted, but that's against the principle of "treating rv as opaque". WDYT?

> If you change the policy to background does the test reliably fail?

Yes, because the rc gets deleted immediately, and gc cannot delete all 100 pods before the `podClient.List()` is issued.

> Does it reliably pass now?

Yes so far.